dutch-auction/src/main.sw

```diff
@@ -0,0 +1,123 @@
+contract;
+
+use std::{
+    address::Address,
+    contract_id::ContractId,
+    block::height,
+    chain::auth::msg_sender,
+    assert::assert,
+    revert::revert,
+    storage::StorageMap,
+    token::transfer_to_output,
+    context::{msg_amount, call_frames::msg_asset_id},
+    result::*,
+    identity::Identity,
+};
+
+abi MyContract {
+    fn get_price() -> u64;
+    fn set_beneficiary(new_beneficiary: Address);
+    fn bid();
+}
+
+//Set this to your own address
+const MY_ADDRESS: b256 = 0x0000000000000000000000000000000000000000000000000000000000000000;
+
+storage {
+    startingPrice: u64,
+    endingPrice: u64,
+    startTime: u64,
+    endTime: u64,
+    beneficiary: Address = ~Address::from(MY_ADDRESS),
+    admin: Address = ~Address::from(MY_ADDRESS),
+    ended: bool,
+    //Enter the asset id of the currency you want the bid to be in
+    asset_id: b256 = 0x0000000000000000000000000000000000000000000000000000000000000000,
+}
+
+fn win() {
+    // Do stuff on the win event
+    transfer_to_output(price(), ~ContractId::from(storage.asset_id), storage.beneficiary);
+    storage.ended = true;
+}
+
+impl MyContract for Contract {
+    fn get_price() -> u64 {
+        return price();
+    }
+
+    fn set_beneficiary(new_beneficiary: Address) {
+        assert(get_sender() == storage.admin);
+
+        storage.beneficiary = new_beneficiary;
+    }
+
+    fn bid() {
+        //Since this is the dutch auction, first bid wins
+        assert(msg_amount() >= price());
+        assert(msg_asset_id() == ~ContractId::from(storage.asset_id));
```

Comment:
When you purchase something it makes sense to first check that you have received the correct thing and then the amount.
For example, if you attempt to purchase something from you for an amount of "5" but I give you 5 carrots that doesn't make since in how you would pick up on that in reality. You wouldn't count the 5 carrots and say "yes, I have expected 5" and then say "wait a minute, these are carrots and not the currency that I expect".
You first notice the currency and then you count the amount